### PR TITLE
Translation function

### LINKED ISSUE #8508: Change function t to translation in app/views/notes/_card.html.erb
Hi, this is a [first-timers-only issue](https://code.publiclab.org/#r=all). This means we've worked to make it more legible to folks who either **haven't contributed to our codebase before, or even folks who haven't contributed to open source before**. 

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

We know that the process of creating a pull request is the biggest barrier for new contributors. This issue is for you 💝

If you have contributed before, **consider leaving this one for someone new**, and looking through our general [help wanted](https://github.com/publiclab/plots2/labels/help-wanted) issues. Thanks!

### 🤔 What you will need to know.

Nothing. This issue is meant to welcome you to Open Source :) We are happy to walk you through the process.

### 📋 Step by Step

- [ ] 🙋 **Claim this issue**: Comment below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!

- [ ] 📝 **Update** the file [app/views/notes/_card.html.erb](https://github.com/publiclab/plots2/blob/master/app/views/notes/_card.html.erb) in the `plots2` repository (press the little pen Icon) and edit the line as shown below. 

[See this page](https://code.publiclab.org/#r=all) for some help in taking your first steps!

Below is a "diff" showing in red (and a `-`) which lines to remove, and in green (and a `+`) which lines to add:

```diff
- 16: <%= t('notes._notes.moderate_first_time_post') %> <br class="d-none d-lg-inline"/>
+ 16: <%= translation('notes._notes.moderate_first_time_post') %> <br class="d-none d-lg-inline"/>

- 17: <a class="btn btn-outline-secondary btn-sm" href="/moderate/publish/<%= node.id %>"><%= t('notes._notes.approve') %></a>
+ 17: <a class="btn btn-outline-secondary btn-sm" href="/moderate/publish/<%= node.id %>"><%= translation('notes._notes.approve') %></a>

- 18: <a class="btn btn-outline-secondary btn-sm" data-test="spam" href="/moderate/spam/<%= node.id %>"><%= t('notes._notes.spam') %></a>
+ 18: <a class="btn btn-outline-secondary btn-sm" data-test="spam" href="/moderate/spam/<%= node.id %>"><%= translation('notes._notes.spam') %></a>

- 20: <%= raw t('notes._notes.pending_approval', :url1 => '/wiki/moderation') %>
+ 20: <%= raw translation('notes._notes.pending_approval', :url1 => '/wiki/moderation') %>

- 49: <li><a>Last Edited: <%= t('notes._notes.last_edit_by') %> <a <% if @widget %>target="_blank"<% end %> href="/profile/<%= node.latest.author.name %>"><%= node.latest.author.name %></a></a></li>
+ 49: <li><a>Last Edited: <%= translation('notes._notes.last_edit_by') %> <a <% if @widget %>target="_blank"<% end %> href="/profile/<%= node.latest.author.name %>"><%= node.latest.author.name %></a></a></li>

- 52: <li><a>Total Views: </i> <%= number_with_delimiter(node.views) %> <span class="d-none d-lg-inline"><%= t('notes._notes.views') %></span></a></li>
+ 52: <li><a>Total Views: </i> <%= number_with_delimiter(node.views) %> <span class="d-none d-lg-inline"><%= translation('notes._notes.views') %></span></a></li>

- 57: <li class="buttons mb-2"><% if @current_user && ['admin','moderator'].include?(@current_user.role) %><a style="width: 90px" class="btn btn-outline-secondary" href="/moderate/spam/<%= node.id %>"><i class="fa fa-ban"></i> <%= t('notes._notes.spam') %></a><% end %></li>
+ 57: <li class="buttons mb-2"><% if @current_user && ['admin','moderator'].include?(@current_user.role) %><a style="width: 90px" class="btn btn-outline-secondary" href="/moderate/spam/<%= node.id %>"><i class="fa fa-ban"></i> <%= translation('notes._notes.spam') %></a><% end %></li>

- 58: <li class="buttons"><a class="btn btn-outline-secondary" href="<%= node.path %>"><%= t('notes._notes.read_more') %></a></li>
+ 58: <li class="buttons"><a class="btn btn-outline-secondary" href="<%= node.path %>"><%= translation('notes._notes.read_more') %></a></li>

- 61: <a class="btn btn-outline-secondary" href="#"><i class="fa fa-ban-circle"></i> <%= t('notes._notes.spam') %></a>
+ 61: <a class="btn btn-outline-secondary" href="#"><i class="fa fa-ban-circle"></i> <%= translation('notes._notes.spam') %></a>
```

- [ ] 💾 **Commit** your changes

- [ ] 🔀 **Start a Pull Request**. There are two ways how you can start a pull request:

1. If you are familiar with the terminal or would like to learn it, [here is a great tutorial](https://egghead.io/series/how-to-contribute-to-an-open-source-project-on-github) on how to send a pull request using the terminal.

2. You can also [edit files directly in your browser](https://help.github.com/articles/editing-files-in-your-repository/) and open a pull request from there. 

- [ ] 🏁 **Done** Ask in comments for a review :)

### Please keep us updated

💬⏰ - We encourage contributors to be respectful to the community and provide an update **within a week** of claiming a first-timers-only issue. We're happy to keep it assigned to you as long as you need if you update us with a request for more time or help, but if we don't see any activity a week after you claim it we may reassign it to give someone else a chance. Thank you in advance! 

If this happens to you, don't sweat it! Grab another open issue.

### Is someone else already working on this?

🔗- We encourage contributors to link to the original issue in their pull request so all users can easily see if someone's already started on it. 

👥- **If someone seems stuck, offer them some help!** Otherwise, [take a look at some other issues you can help with](https://code.publiclab.org/#r=all). Thanks!

### 🤔❓ Questions?

Leave a comment below!

### PR DESCRIPTION
**Pull request by Priyaraj17**
Fixes #8508 (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts 📁
* [ ] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

I have updated the _card.html.erb file and replaced the function t with translation.
> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Review by Tlazypanda — COMMENTED** [review]
@Priyaraj17 can you remove the files `_header.html.erb` and `methods.html.erb` changes? :sweat_smile:

**Comment by Priyaraj17** [comment]
Sorry, I didn't see your message. Sure I will remove them.

**Comment by Priyaraj17** [comment]
> @Priyaraj17 can you remove the files `_header.html.erb` and `methods.html.erb` changes?

Hey @Tlazypanda, sorry, I was working on my proposal. Can you please tell me what do I have to remove? I believe these are the changes I made when I was working on a different pull request.

**Comment by Priyaraj17** [comment]
> > @Priyaraj17 can you remove the files `_header.html.erb` and `methods.html.erb` changes?
> 
> Hey @Tlazypanda, sorry, I was working on my proposal. Can you please tell me what do I have to remove? These are the changes I made when I was working on a different issue.

File: app/views/wiki/_header.html.erb [hunk 1]
```diff
@@ -31,7 +31,13 @@
...
     <div style="padding-top:8px;text-align:center">
       <span class="d-none d-lg-inline">
-        <span rel="tooltip" title="The number of authors for this page." data-placement="top"><i style="color:#888;" class="fa fa-user"></i> <%= number_with_delimiter(@node.authors.size) %></span>  |
+        <span rel="popover" style='cursor: pointer;' title="Authors" data-html="true" data-placement="top" data-content= "
+        <ul>
+        <% @node.authors.each do |author| %>
+          <li style='list-style: none;'><a href='/profile/<%= author.username %>' style='color: #009bff;'>@<%= author.username %></a></li>
+        <% end %>
+      </ul>
+      "> <i style="color:#888;" class="fa fa-user"></i> <%= number_with_delimiter(@node.authors.size) %></span>  |
```
Comment (RuthNjeri) [inline app/views/wiki/_header.html.erb:40]: @Priyaraj17, this is the file @Tlazypanda was mentioning. These changes are not required.

File: app/views/wiki/_header.html.erb [hunk 2]
```diff
@@ -62,3 +68,5 @@
   </ul>
 </nav>
 <% end %>
+
+
```
Comment (RuthNjeri) [inline app/views/wiki/_header.html.erb:72]: These changes are not needed as well

File: app/views/wiki/methods.html.erb [hunk 3]
```diff
@@ -62,3 +62,4 @@
   </div>
   <hr style="clear:both;" />
 </div>
+e
```
Comment (RuthNjeri) [inline app/views/wiki/methods.html.erb:65]: @Priyaraj17, this is the second file @Tlazypanda was mentioning. This change is not required.

Comment (Priyaraj17) [reply app/views/wiki/methods.html.erb]: Hey, I have removed the changes.

**Comment by gaurav2699** [comment]
Hey @Priyaraj17 there are still some unnecessary changes in the ```_header.html.erb```  and ```methods.html.erb``` files due to a different PR made by you. Try to use different branches to avoid this issue in the future. To fix this issue you can do ```git checkout staging -- <file_to_be_removed>``` for the two files to remove those files from the remote. This should fix the issue and your PR would be good to merge. Thanks!

**Comment by gaurav2699** [comment]
Hey @Priyaraj17, there are still unnecessary changes in the ```_header.html.erb``` and ```methods.html.erb``` files. This is also causing a merge conflict.

**Comment by Priyaraj17** [comment]
> Hey @Priyaraj17, there are still unnecessary changes in the `_header.html.erb` and `methods.html.erb` files. This is also causing a merge conflict.

Hi, I think the problem appears because of an earlier PR, I made. I am not sure what to do now.

**Comment by gaurav2699** [comment]
A simple solution for this can be, just check out the file changed section on top of this PR, where you will be shown all changes you made in all the files. As you can see, it shows the other two files as changed and the changes you made in those files. Your job is to just revert the changes in the two file, so you just copy the red part and replace it with the green part. 
Another pro solution for this could be:
``` git fetch upstream ```
``` git checkout upstream/main -- app/views/wiki/_header.html.erb ```
``` git checkout upstream/main -- app/views/wiki/methods.html.erb```
After that, just commit and push. Hopefully, this fixes the issue. This just reverts those files back to the publiclab version in the remote. Thanks!

**Comment by govindgoel** [comment]
@Priyaraj17 , 
@gaurav2699 has mentioned both the ways that can solve the issue, you just have to remove the changes from the above mentioned files and then update your PR branch.

**Comment by Priyaraj17** [comment]
> A simple solution for this can be, just check out the file changed section on top of this PR, where you will be shown all changes you made in all the files. As you can see, it shows the other two files as changed and the changes you made in those files. Your job is to just revert the changes in the two file, so you just copy the red part and replace the green part with it.
> Another pro solution for this could be:
> `git fetch upstream`
> `git checkout upstream/main -- app/views/wiki/_header.html.erb`
> ` git checkout upstream/main -- app/views/wiki/methods.html.erb`
> After that, just commit and push. Hopefully, this fixes the issue. This just reverts those files back to the publiclab version in the remote. Thanks!

Hey @gaurav2699 , Thanks for helping me out. I hope this time the PR won't have any conflicts.

**Comment by Priyaraj17** [comment]
> @Priyaraj17 ,
> @gaurav2699 has mentioned both the ways that can solve the issue, you just have to remove the changes from the above mentioned files and then update your PR branch.

Yeah, Thanks...for the help. I have changed all the things.

**Comment by gaurav2699** [comment]
Good job! You fixed the unnecessary changes in the  ```methods.html.erb``` file, so one down. Although, I see there are still changes in ```_header.html.erb``` file. Can you fix that in the same way?

File: app/views/wiki/_header.html.erb (hunk 1 above)
Comment (gaurav2699) [inline app/views/wiki/_header.html.erb:40]: Revert these changes basically.

**Review by gaurav2699 — CHANGES_REQUESTED** [review]

File: app/views/wiki/_header.html.erb (hunk 1 above)
Comment (Priyaraj17) [reply app/views/wiki/_header.html.erb:40]: I thought they were changed. Anyways I will change them now. Thanks

Comment (Priyaraj17) [reply app/views/wiki/_header.html.erb]: @gaurav2699 Hey, I have reverted the changes, I hope this might work. Sorry about this. It's a very simple contribution and its taking so time.

File: app/views/wiki/_header.html.erb [hunk 4]
```diff
@@ -31,7 +31,7 @@
 
     <div style="padding-top:8px;text-align:center">
       <span class="d-none d-lg-inline">
-        <span rel="tooltip" title="The number of authors for this page." data-placement="top"><i style="color:#888;" class="fa fa-user"></i> <%= number_with_delimiter(@node.authors.size) %></span>  |
+      <span rel="tooltip" title="The number of authors for this page." data-placement="top"><i style="color:#888;" class="fa fa-user"></i> <%= number_with_delimiter(@node.authors.size) %></span> |
```
Comment (gaurav2699) [inline app/views/wiki/_header.html.erb:34]: Good job, we are very close, but there is still a couple of things, like indent this line of code. Git detects every small change.

File: app/views/wiki/_header.html.erb [hunk 5]
```diff
@@ -62,3 +62,5 @@
   </ul>
 </nav>
 <% end %>
+
+
```
Comment (gaurav2699) [inline app/views/wiki/_header.html.erb:66]: Also remove these two lines, then this PR will be good to merge.

File: app/views/wiki/_header.html.erb (hunk 4 above)
Comment (Priyaraj17) [reply app/views/wiki/_header.html.erb:34]: Hey, thanks Infact I was thinking that this indentation might create a problem but wasn't sure. But thanks I will change it now.

**Comment by govindgoel** [comment]
@Priyaraj17 There is still a conflict in this file app/views/wiki/_header.html.erb you need to resolve that

**Comment by RuthNjeri** [comment]
Thanks, @gaurav2699 and @govindgoel 

@Priyaraj17 you can refer to this comment https://github.com/publiclab/plots2/pull/9491#issuecomment-821788391 in fixing the current merge conflict.

**Comment by gaurav2699** [comment]
@Priyaraj17 Please run the following commands in your command line of the root directory:
```git fetch upstream```
```git checkout upstream/main -- app/views/wiki/_header.html.erb```
Then commit and push.

**Comment by govindgoel** [comment]
@Priyaraj17 A quick fix for this, I would suggest as there are around 15 commits in your branch and having different changes. What you can do is remove these commits from your local then see `git status` if everything is fine do the necessary changes and check with `git diff` and push that single commit with the changes. And it would only reflect changes in the files you edited.
Thanks

**Comment by Priyaraj17** [comment]
> @Priyaraj17 A quick fix for this, I would suggest as there are around 15 commits in your branch and having different changes. What you can do is remove these commits from your local then see `git status` if everything is fine do the necessary changes and check with `git diff` and push that single commit with the changes. And it would only reflect changes in the files you edited.
> Thanks

Thanks for this. It worked.

File: app/views/wiki/_header.html.erb (hunk 1 above)
Comment (jywarren) [reply app/views/wiki/_header.html.erb]: That's OK @Priyaraj17 it's always a learning process and we appreciate your time too!

**Comment by jywarren** [comment]
This looks good, but i'm not sure why the tests are failing... Thanks everyone for pitching in here, we are almost there! I'll look at the test output...

**Comment by jywarren** [comment]
Wow hmm, on both i see:

```

Minitest::UnexpectedError:         ActionView::Template::Error: undefined method `translation' for #<#<Class:0x00007fb20b082978>:0x00007fb20b013690>
            app/views/notes/_card.html.erb:54:in `_app_views_notes__card_html_erb___1103013193437718487_70201334299660'
            app/views/notes/_notes.html.erb:9:in `block in _app_views_notes__notes_html_erb__4411833217512531815_70201334250280'
            app/views/notes/_notes.html.erb:6:in `each'
            app/views/notes/_notes.html.erb:6:in `each_with_index'
            app/views/notes/_notes.html.erb:6:in `_app_views_notes__notes_html_erb__4411833217512531815_70201334250280'
            app/views/grids/_thumbnail.html.erb:2:in `_app_views_grids__thumbnail_html_erb__3613994063507155560_70201334220080'
            app/models/concerns/node_shared.rb:345:in `data_string'
            app/models/concerns/node_shared.rb:34:in `block in notes_thumbnail_grid'
            app/models/concerns/node_shared.rb:21:in `gsub'
            app/models/concerns/node_shared.rb:21:in `notes_thumbnail_grid'
            app/helpers/application_helper.rb:89:in `insert_extras'
            app/views/wiki/show.html.erb:41:in `_app_views_wiki_show_html_erb___1463884476226702321_70201215479520'
```

I wonder why the translation helper isn't found???

**Comment by jywarren** [comment]
ahhh, i bet... i bet it's because we're trying to render something inside a model/concern. Things are a little different in there. Let me see...

File: app/views/notes/_card.html.erb [hunk 6]
```diff
@@ -13,11 +13,11 @@
     <% if node.status == 4 %>
       <p class="alert alert-warning moderated">
         <% if @current_user && ['admin','moderator'].include?(@current_user.role) %>
-          <%= t('notes._notes.moderate_first_time_post') %> <br class="d-none d-lg-inline"/>
-          <a class="btn btn-outline-secondary btn-sm" href="/moderate/publish/<%= node.id %>"><%= t('notes._notes.approve') %></a>
-          <a class="btn btn-outline-secondary btn-sm" data-test="spam" href="/moderate/spam/<%= node.id %>"><%= t('notes._notes.spam') %></a>
+          <%= translation('notes._notes.moderate_first_time_post') %> <br class="d-none d-lg-inline"/>
```
Comment (jywarren) [inline app/views/notes/_card.html.erb:16]: OK, could we try this? 

```suggestion
          <%= ApplicationController.helpers.translation('notes._notes.moderate_first_time_post') %> <br class="d-none d-lg-inline"/>
```
And repeat that change for all instances of `translation()` in this file? 

What i think is happening is that this template file is (uniquely, we don't do this for other templates) being rendered inside the `/app/models/concerns/node_shared.rb` file as we've modularized the generation of this code a bit. And in model or concern code, we don't have "quick access" to view helpers. We need instead to use a full reference with `ApplicationController.helpers._______` to get to them. 

Let's make this change throughout this one file and see if it passes tests then!

**Review by jywarren — CHANGES_REQUESTED** [review]

File: app/views/notes/_card.html.erb (hunk 6 above)
Comment (Priyaraj17) [reply app/views/notes/_card.html.erb:16]: Okay, got it. Shall I make the changes?

Comment (RuthNjeri) [reply app/views/notes/_card.html.erb:16]: Yes @Priyaraj17, please make the changes...

Comment (Priyaraj17) [reply app/views/notes/_card.html.erb]: Hey, @RuthNjeri  @jywarren 
I have made the changes. I hope this will be fine.

Comment (jywarren) [reply app/views/notes/_card.html.erb]: Oof, that didn't work, did it? Gosh, sorry @priyaraj17 - i think we need to look at the error logs and potentially look at other options for how to use the helpers:

https://stackoverflow.com/questions/6820547/use-a-rails-helper-in-a-model

Comment (Priyaraj17) [reply app/views/notes/_card.html.erb]: Yeah, It didn't work. In fact, I created a PR in the public lab editor repo a couple of weeks ago. A few tests were failing in that PR too.

Comment (RuthNjeri) [reply app/views/notes/_card.html.erb]: Should we also check if `current_user` exists?

**Comment by Priyaraj17** [comment]
Hey @cesswairimu @jywarren what should I do here?

**Comment by RuthNjeri** [comment]
@Priyaraj17, try updating the translation method in `app/helpers/application_helper.rb` to 

```diff
def translation(key, options = {})
    translated_string = t(key, options)
    options[:fallback] = false
    translated_string2 = t(key, options)
+    # To prevent no method error when app/models/concerns/node_shared.rb calls app/views/notes/_card.html.erb
+    # which then calls this method without current_user
+   begin
      if current_user&.has_tag('translation-helper') && translated_string2.include?("translation missing") && !translated_string.include?("<")
        raw(%(<span>#{translated_string} <a href="https://www.transifex.com/publiclab/publiclaborg/translate/#de/$?q=text%3A#{translated_string}">
            <i data-toggle='tooltip' data-placement='top' title='Needs translation? Click to help translate this text.' style='position:relative; right:2px; color:#bbb; font-size: 15px;' class='fa fa-globe'></i></a>
         </span>))
      else
        raw(translated_string)
      end
+    rescue
      raw(translated_string)
+    end
  end
```

**Comment by Priyaraj17** [comment]
Okay, @RuthNjeri Thanks!

**Comment by Priyaraj17** [comment]
Hey @RuthNjeri, Sorry for being this late, I have made the changes as mentioned by you. This is the [PR](https://github.com/publiclab/plots2/pull/9669), I created a different branch to make the changes

**Comment by RuthNjeri** [comment]
Thanks, @Priyaraj17, I have made a suggestion in the other PR... I'm curious why a second branch was created? Can the changes be made here? That way even old discussions can be referenced by new reviewers.

**Comment by Priyaraj17** [comment]
> Thanks, @Priyaraj17, I have made a suggestion in the other PR... I'm curious why a second branch was created? Can the changes be made here? That way even old discussions can be referenced by new reviewers.

Hey, sorry didn't see your comment.  I was working on a different issue earlier with a different branch, I will update the changes. Thanks!
